runtime/caml/misc.h

```diff
@@ -215,13 +215,9 @@ Caml_inline void call_timing_hook(_Atomic caml_timing_hook * a)
 
 #endif /* CAML_INTERNALS */
 
-#define CAML_STATIC_ASSERT_3(b, l) \
-  CAMLunused_start \
-    CAMLextern char static_assertion_failure_line_##l[(b) ? 1 : -1] \
-  CAMLunused_end
-
-#define CAML_STATIC_ASSERT_2(b, l) CAML_STATIC_ASSERT_3(b, l)
-#define CAML_STATIC_ASSERT(b) CAML_STATIC_ASSERT_2(b, __LINE__)
+#if defined (__STDC_VERSION__) && __STDC_VERSION__ < 202311L
+#include <assert.h> /* for the static_assert macro */
+#endif
```

Comment:
We've had problems in the past with including `<stdbool.h>` from `<caml/misc.h>`, as some 3rd party C code was defining their own `bool` type, conflicting with the one defined in `<stdbool.h>`.  So, to be very prudent, I would remove these 3 lines, and add `#include <assert.h>` (unconditionally) in the .c files that need `static_assert`.  (Unconditionally, because in C23 the include is not needed but harmless.)

There remains the `static_assert` in `<caml/domain_state.h>`.  Adding `#include <assert.h>` there is about as bad as adding it to `<caml/misc.h>`.  In perfect hindsight, I'd rather have this assertion in a C file, e.g. runtime/domain.c, or protect it with a `#ifdef CAML_INTERNALS`, so that no namespace pollution occurs for 3rd-party code.  @kayceesrk , what do you think about this?

Reply:
Protecting the assertion with `#ifdef CAML_INTERNALS` is the easier option. I would be ok with that change.

Reply:
I've moved the assertions to C files. Maybe they're less clear since they're not located next to what they're testing?

Reply:
This is ok to me. This is not something that we expect to fail during casual hacking. You should know what you are doing when you add fields to Caml_state.